# Denison CS-181/DA-210 Homework

---

## Combining Tables Exercises

In [2]:
import os
import io
import sys
import re
import pandas as pd

from contextlib import redirect_stdout

def add_modules():
    """
    Starting at the current directory and proceeding up the file system
    tree, search for a directory named `modules`.  If found, and if not
    already there, add to the Python module search path.
    
    Params: None
    
    Return: None
    """
    directory = "."
    levels = 0
    while not os.path.isdir(os.path.join(directory, "modules")) and \
          levels < 5:
        directory = os.path.join(directory, "..")
        levels += 1
    module_path = os.path.abspath(os.path.join(directory, "modules"))
    if os.path.isdir(module_path):
        if not module_path in sys.path:
            sys.path.append(module_path)

add_modules()
import util

datadir = util.resolve_dir("tabulardata")

**Q1** In the data directory, you will find two csv files, `educationTop.csv` and `educationBottom.csv`, both based on data hosted by `www.census.gov`. Both have the same columns, and each row is a U.S. metropolitan area, with data on population, education, and unemployment. The first `csv` file contains metropolitan areas starting with A-K, and the second starting with L-Z. Read both into `pandas` data frames, using the column `GEO.id2` as an index. Be careful, as these files have **two** lines that might be considered header lines.  The first one is more descriptive, and the second one is more concise and gives succint variable names.  So when we read from the csv, we want to **skip** the first line and then use the second line as the header line.  The `pandas` `read_csv` constructor can do this for you, but you have to discover and use the correct arguments.

Concatenate these two data frames along the row dimension (with the top one on top), and call the result `educationDF`.

In [ ]:
# Student Experimentation Cell



In [13]:
# Solution Cell
educationTop = pd.read_csv(os.path.join(datadir,'educationTop.csv'), skiprows=1, index_col='GEO.id2')
educationBot = pd.read_csv(os.path.join(datadir,'educationBottom.csv'), skiprows=1, index_col='GEO.id2')
educationDF = pd.concat([educationTop, educationBot], axis=0)
educationDF.head()

GEO.id                       GEO.display-label  HD01_VD01  \
GEO.id2                                                                      
10180    310M100US10180                  Abilene, TX Metro Area      82270   
10420    310M100US10420                    Akron, OH Metro Area     369311   
10500    310M100US10500                   Albany, GA Metro Area      78159   
10580    310M100US10580  Albany-Schenectady-Troy, NY Metro Area     466236   
10740    310M100US10740              Albuquerque, NM Metro Area     476942   

         HD02_VD01  HD01_VD02  HD02_VD02  HD01_VD03  HD02_VD03  HD01_VD04  \
GEO.id2                                                                     
10180         1582      13820       1873       6484       1272          0   
10420         1080      26613       2972      14519       2048          0   
10500         1072      13689       2238       7681       1645          0   
10580         1818      29031       3211      16906       2415          0   
10740         1656      51438       3778      30718       3015          0   

         HD02_VD04  ...  HD01_VD25  HD02_VD25  HD01_VD26  HD02_VD26  \
GEO.id2             ...                                               
10180          206  ...        579        347      12909       1564   
10420          182  ...         43         70      97693       4241   
10500          208  ...        247        249      12051       1611   
10580          185  ...        572        510     146172       5083   
10740          185  ...       1179        623     118976       5222   

         HD01_VD27  HD02_VD27  HD01_VD28  HD02_VD28  HD01_VD29  HD02_VD29  
GEO.id2                                                                    
10180        12770       1557        139        137       3633        826  
10420        94492       4200       3201        888      13895       1627  
10500        11493       1472        558        507       2599        678  
10580       142431       5036       3741        887      22891       1945  
10740       114318       5071       4658        986      21356       2336  

[5 rows x 60 columns]

In [ ]:
# Testing Cell

assert True

**Q2** In the data directory, you will find two csv files, `educationLeft.csv` and `educationRight.csv`, both based on data hosted by `www.census.gov`. Both have the same rows, and each row is a U.S. metropolitan area, with data on population, education, and unemployment. The first has information on individuals without a college degree, and the second has information on individuals with a college degree. Read both into `pandas` data frames. The `GEO.display-label` field should serve as the row index.  Read these files into data frames and then concatenate these two data frames along the column dimension, and call the result `educationDF2`.

The same cautions on the CSV having an informational row of headers followed by the header variable names applies to these two CSV files as well.

In [ ]:
# Student Experimentation Cell



In [20]:
# Solution Cell
educationLeft = pd.read_csv(os.path.join(datadir, "educationLeft.csv"), index_col= 'GEO.display-label', skiprows=1)
educationRight = pd.read_csv(os.path.join(datadir, "educationRight.csv"), index_col= 'GEO.display-label',skiprows=1)
educationDF2 = pd.concat([educationLeft, educationRight], axis=1)
educationDF2.head()

HD01_VD01  HD02_VD01  HD01_VD02  \
GEO.display-label                                                         
Abilene, TX Metro Area                      82270       1582      13820   
Akron, OH Metro Area                       369311       1080      26613   
Albany, GA Metro Area                       78159       1072      13689   
Albany-Schenectady-Troy, NY Metro Area     466236       1818      29031   
Albuquerque, NM Metro Area                 476942       1656      51438   

                                        HD02_VD02  HD01_VD03  HD02_VD03  \
GEO.display-label                                                         
Abilene, TX Metro Area                       1873       6484       1272   
Akron, OH Metro Area                         2972      14519       2048   
Albany, GA Metro Area                        2238       7681       1645   
Albany-Schenectady-Troy, NY Metro Area       3211      16906       2415   
Albuquerque, NM Metro Area                   3778      30718       3015   

                                        HD01_VD04  HD02_VD04  HD01_VD05  \
GEO.display-label                                                         
Abilene, TX Metro Area                          0        206       6484   
Akron, OH Metro Area                            0        182      14519   
Albany, GA Metro Area                           0        208       7681   
Albany-Schenectady-Troy, NY Metro Area          0        185      16906   
Albuquerque, NM Metro Area                      0        185      30718   

                                        HD02_VD05  ...  HD01_VD25  HD02_VD25  \
GEO.display-label                                  ...                         
Abilene, TX Metro Area                       1272  ...        579        347   
Akron, OH Metro Area                         2048  ...         43         70   
Albany, GA Metro Area                        1645  ...        247        249   
Albany-Schenectady-Troy, NY Metro Area       2415  ...        572        510   
Albuquerque, NM Metro Area                   3015  ...       1179        623   

                                        HD01_VD26  HD02_VD26  HD01_VD27  \
GEO.display-label                                                         
Abilene, TX Metro Area                      12909       1564      12770   
Akron, OH Metro Area                        97693       4241      94492   
Albany, GA Metro Area                       12051       1611      11493   
Albany-Schenectady-Troy, NY Metro Area     146172       5083     142431   
Albuquerque, NM Metro Area                 118976       5222     114318   

                                        HD02_VD27  HD01_VD28  HD02_VD28  \
GEO.display-label                                                         
Abilene, TX Metro Area                       1557        139        137   
Akron, OH Metro Area                         4200       3201        888   
Albany, GA Metro Area                        1472        558        507   
Albany-Schenectady-Troy, NY Metro Area       5036       3741        887   
Albuquerque, NM Metro Area                   5071       4658        986   

                                        HD01_VD29  HD02_VD29  
GEO.display-label                                             
Abilene, TX Metro Area                       3633        826  
Akron, OH Metro Area                        13895       1627  
Albany, GA Metro Area                        2599        678  
Albany-Schenectady-Troy, NY Metro Area      22891       1945  
Albuquerque, NM Metro Area                  21356       2336  

[5 rows x 58 columns]

In [21]:
# Testing Cell

assert True

**Q3** In the data directory, you will find two csv files, `educationLeftJ.csv` and `educationRightJ.csv`, both based on data hosted by `www.census.gov`. In both, rows represent U.S. metropolitan area, with data on population, education, and unemployment. However, they do not have exactly the same set of rows, and the columns are totally different except for the index column `Geography`. In these CSV files, there is not an "extra" header line.  The row label index should come from `Geography` unless you are instructed otherwise.

0. Read both into `pandas` data frames, with names `educationLeftJ` and `educationRightJ`.
1. Make a copy of `educationLeftJ` called `educationLeftOriginal`.
2. Starting with `educationLeftJ`, do a left join to bring in the data from `educationRightJ`, storing your answer as `dfJ`.
3. Starting with `educationLeftOriginal`, do an inner join to bring in the data from `educationRightJ`, storing your answer as `dfJ2`.
4. Now read the original csv files in as `eduLeft` and `eduRight` with no meaningful index. Then, starting from `eduLeft`, do an inner merge along the column `Geography`, storing your answer as `dfJ3`.

In [ ]:
# Student Experimentation Cell



In [22]:
# Solution Cell
educationLeftJ = pd.read_csv(os.path.join(datadir, "educationLeftJ.csv"), index_col = "Geography")
educationRightJ = pd.read_csv(os.path.join(datadir, "educationRightJ.csv"), index_col= "Geography")

educationLeftOriginal = educationLeftJ.copy()

dfJ = educationLeftJ.join(educationRightJ, how="left")

dfJ2 = educationLeftOriginal.join(educationRightJ, how="inner")

eduLeft = pd.read_csv(os.path.join(datadir, "educationLeftJ.csv"))
eduRight = pd.read_csv(os.path.join(datadir, "educationRightJ.csv"))
dfJ3 = pd.merge(eduLeft, eduRight, on='Geography', how='inner')
print(dfJ.shape)
print(dfJ2.shape)
print(dfJ3.shape)

(249, 24)
(152, 24)
(152, 25)


In [23]:
# Testing Cell

assert True

**Q4** Explain the difference in the number of rows and columns in `dfJ`, `dfJ2`, and `dfJ3`.  Be specific.

We can see that educationLeftJ has more rows than educationRightJ because when merging left we have more rows than when doing an inner merge. So we know that educationLeftJ has 249 rows while they both share 152 rows.

**Q5** In the tabular data directory, there are csv files `students1.csv` and `majors.csv`.  Both are part of a multi-table data set we will use in our exploration of the relational data model.  The data set is based on a one year snapshot from Denison from 10 years ago, and then anonymized.  The `students1.csv` is a random selection of 50 students from the 2200 students in the full `students.csv`.

Examine both of these tables using a spreadsheet type view and take a little time to understand the data (values, variables, and which of the vars are independent and dependent).

The students table has information about students, including the up-to-four-character id indicating the major of the student, while the majors table relates the major id to the dependent variables of the full text name of the major and the up-to-four-character id of the *department* of the major.

Suppose we want to compose a table, **sorted by last name, then first name**, of the students, and containing the columns: last name, first name, ethnicity, major id, department id, and major name, in that column order, and with an index of the studentid.  The table should be referred to by Python variable `studentmajors` when complete.  Note that some students will have missing data for their major (i.e. they have not selected a major yet), and that is ok, and they should not "disappear" from the composed table.

Unlike the prior problems where the steps were pretty obvious from the set up, here you need to think about the "right" operation(s) and order of steps needed to combine the tables, order the columns appropriately, sort the rows appropriately, and to have the correct index.

There will be multiple correct ways of solving this problem, and you are welcome to find the way that makes the most sense to you.  I strongly suggest that you work incrementally, and check your results after each step.

In [71]:
# Student Experimentation Cell

students

studentid studentfirst studentlast  studentcity studentstate  \
0         61500         Gary      Ortega          NaN          NaN   
1         61501        Ayaan       Sarin       Newark           OH   
2         61502        Jacob     Hoffman    Granville           OH   
3         61503         Sara        Hart    Granville           OH   
4         61504       Harini       Kapur    Pataskala           OH   
...         ...          ...         ...          ...          ...   
2269      63769      Zachary     Hawkins  Los Angeles           CA   
2270      63770        James    Castillo    Granville           OH   
2271      63771     Aaradhya    Dasgupta        Heath           OH   
2272      63772        Chloe    Campbell       Hebron           OH   
2273      63773       Sophie        Shaw    Granville           OH   

     studentgender studentethnicity majorid  
0                M         Multiple     NaN  
1                M            White     NaN  
2                M            White     NaN  
3                F            White     NaN  
4                F            White     NaN  
...            ...              ...     ...  
2269             M            White    ENGW  
2270             M            White     NaN  
2271             F            White     NaN  
2272             F            White     NaN  
2273             F            Black     NaN  

[2274 rows x 8 columns]

In [75]:
# Solution Cell
students = pd.read_csv(os.path.join(datadir, 'students.csv'))
majors = pd.read_csv(os.path.join(datadir, 'majors.csv'))

studentmajors = pd.merge(students[['studentlast','studentfirst','studentethnicity','majorid']],majors[['majorid','departmentid','majorname']],how='left').set_index(['studentlast','studentfirst'])

print(studentmajors.head())
print(studentmajors.tail())

                         studentethnicity majorid departmentid majorname
studentlast studentfirst                                                
Ortega      Gary                 Multiple     NaN          NaN       NaN
Sarin       Ayaan                   White     NaN          NaN       NaN
Hoffman     Jacob                   White     NaN          NaN       NaN
Hart        Sara                    White     NaN          NaN       NaN
Kapur       Harini                  White     NaN          NaN       NaN
                         studentethnicity majorid departmentid  \
studentlast studentfirst                                         
Hawkins     Zachary                 White    ENGW         ENGL   
Castillo    James                   White     NaN          NaN   
Dasgupta    Aaradhya                White     NaN          NaN   
Campbell    Chloe                   White     NaN          NaN   
Shaw        Sophie                  Black     NaN          NaN   

                          

In [ ]:
# Testing Cell

assert True
